### PR TITLE
Apack dependencies

### PR DESCRIPTION
**Pull request by jrodriguez-rc**
Support APACK dependencies
* Check installed apack package dependencies for the current repository
* Display a popup when not met dependencies

Testing repositories:
https://github.com/jrodriguez-abapGit-Testing/package1
https://github.com/jrodriguez-abapGit-Testing/package2
https://github.com/jrodriguez-abapGit-Testing/package3

### CODE REVIEW

**Comment by larshp** [comment]
A: is it possible not to store the `dot_apack` in database? its used at deserialization only? I dont remember exactly

B: UI decoupling,
`METHOD deserialize_checks,` should not show any gui, but just calulate the status

it is then up to 
https://github.com/larshp/abapGit/blob/714fe7202dbd225433a42dfe08f66753e6f5005b/src/ui/zcl_abapgit_services_repo.clas.abap#L93 to present the user interface

**Comment by larshp** [comment]
@SebastianWolf-SAP do you have time to take a look?

**Comment by jrodriguez-rc** [comment]
Ok, I tried to do the same that .abapgit file and requirement checks. I'll change to only deserialize .apack file and perform checks and move popup to `gui_deserialize`

**Comment by jrodriguez-rc** [comment]
Testing repositories:
https://github.com/jrodriguez-abapGit-Testing/package1
https://github.com/jrodriguez-abapGit-Testing/package2

**Comment by SebastianWolf-SAP** [comment]
Wow, cool. Thanks, @jrodriguez-rc !

Unfortunately, I won't have time before next week to perform an extensive testing. But I will definitely do so. 

For now, I only have two remarks: 
* One of the kind "nitpicking": I saw that there is a typo in the importing parameter it_dependecies  - an 'n' is missing...
* One more general thing: I think the Git URL should not be part of the check if the dependency is met or not. If I recall correctly, this was probably (or is still) wrong in the original implementation, but in line with a major revamp while introducing versioned dependencies, I only compare group and artifact ID, see in my fork here: https://github.com/SebastianWolf-SAP/ADT_Frontend/blob/master/org.abapgit.adt.ui/src/org/abapgit/adt/ui/internal/wizards/AbapGitWizardPageBranchAndPackage.java#L447 Unfortunately, I haven't had a chance to push the latest changes as I wanted to introduce PDE tests along with it. Maybe I urgently need to free some time to continue here...

As said, more details next week when I can do a detailed test. If you would like to talk beforehand, let me know. There is always time for a short meeting. ;)

**Comment by jrodriguez-rc** [comment]
Copy and paste are the worst for typo mistakes, thanks for it.

I remove repository address from the check, I included it just in case someone uses the same group and artifact.

Just to know, is there any reason that the version is not included in the dependency?
It can be important, like software component release and patch

**Comment by SebastianWolf-SAP** [comment]
The reason is simple: I haven't had time to do that in the first round of the implementation. I also consider it to be very important, but I wanted to ship a minimum viable product and then extend it step by step. 

The code which you can see in my fork on https://github.com/SebastianWolf-SAP/ADT_Frontend already includes these version dependency management, but due to some reorgs and continuous workload I there was no time so far to complete it (PDE tests missing). Please keep the fingers crossed that there will be time to complete it soon.

**Comment by jrodriguez-rc** [comment]
Ah ok, thank you for the answer.

If it is in progress for ADT Frontend, and if it is ok for you, I will implement version check in this PR

**Comment by SebastianWolf-SAP** [comment]
Absolutely OK for me, thanks already! :)

**Comment by jrodriguez-rc** [comment]
Done:
image

- Green: Dependency OK
- Grey: Dependency installed but in lower version
- Red: Dependency not installed

File: src/apack/zcl_abapgit_apack_helper.clas.abap [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+                 artifact_id = ls_dependecy-artifact_id.
+      IF sy-subrc <> 0.
+        ls_dependecy_popup-met = 'N'.
+      ELSE.
+        TRY.
+            zcl_abapgit_version=>check_dependant_version( is_current   = ls_installed_package-sem_version
+                                                          is_dependant = ls_dependecy-sem_version ).
+            ls_dependecy_popup-met = 'Y'.
+          CATCH zcx_abapgit_exception.
+            ls_dependecy_popup-met = 'P'.
```
Comment (larshp) [inline src/apack/zcl_abapgit_apack_helper.clas.abap:129]: suggest adding constants for these

> met = 'N'.
> met = 'Y'.
> met = 'P'.

File: src/apack/zcl_abapgit_apack_reader.clas.abap [hunk 2]
```diff
@@ -56,6 +83,41 @@ CLASS zcl_abapgit_apack_reader IMPLEMENTATION.
...
+
+  METHOD from_xml.
+
+    DATA: lv_xml TYPE string.
+
+    lv_xml = iv_xml.
+
+    " fix downward compatibility
+    REPLACE ALL OCCURRENCES OF '<_--28C_DATA_--29>' IN lv_xml WITH '<DATA>'.
+    REPLACE ALL OCCURRENCES OF '</_--28C_DATA_--29>' IN lv_xml WITH '</DATA>'.
```
Comment (larshp) [inline src/apack/zcl_abapgit_apack_reader.clas.abap:111]: these 2 statements can be removed, its for old stuff, before apack was introduced, so the apack xmls should be okay

File: src/zcl_abapgit_repo.clas.abap [hunk 3]
```diff
@@ -56,6 +56,14 @@ CLASS zcl_abapgit_repo DEFINITION
         !io_dot_abapgit TYPE REF TO zcl_abapgit_dot_abapgit
       RAISING
         zcx_abapgit_exception .
+    METHODS get_dot_apack
+      RETURNING
+        VALUE(ro_dot_apack) TYPE REF TO zcl_abapgit_apack_reader .
+    METHODS set_dot_apack
```
Comment (larshp) [inline src/zcl_abapgit_repo.clas.abap:62]: this method can be moved to private?

File: src/apack/zcl_abapgit_apack_helper.clas.abap (hunk 1 above)
Comment (jrodriguez-rc) [reply src/apack/zcl_abapgit_apack_helper.clas.abap:129]: Ok, should I use those constants also for abapGit requirements Y/N?

File: src/zcl_abapgit_repo.clas.abap (hunk 3 above)
Comment (jrodriguez-rc) [reply src/zcl_abapgit_repo.clas.abap:62]: I can move `set_dot_apack`, but `get_dot_apack` is used from `zcl_abapgit_services_repo`

Comment (larshp) [reply src/zcl_abapgit_repo.clas.abap:62]: yes, only set_dot_apack should be set to private

File: src/apack/zcl_abapgit_apack_helper.clas.abap (hunk 1 above)
Comment (larshp) [reply src/apack/zcl_abapgit_apack_helper.clas.abap:129]: ideally, yes. But its probably easier to leave all this as-is, both old and new code, and then defer it to a future issue/PR

File: src/zcl_abapgit_repo.clas.abap (hunk 3 above)
Comment (larshp) [reply src/zcl_abapgit_repo.clas.abap]: done in https://github.com/larshp/abapGit/pull/3285/commits/85566af6d3a0b8d286dca995ab6a2a9db987ac56

**Comment by SebastianWolf-SAP** [comment]
Now that I finally have time, the PR is already merged. Well, I would have given green light anyway, nice solution with the popup. I was still thinking that I need to include the complete automated mass import, but that is really a nice solution.

Thanks @larshp for the review and merge. And of course, thanks again @jrodriguez-rc for the functionality! Glad to see that there is progress.

**Comment by larshp** [comment]
Did not review it in super detail, most of this is new isolated functionality..

Feel free to open issues for follow up changes

**Pull request merged**